File: week1_baseline/python/src/boukensha/prompt_builder.py
from __future__ import annotations


class PromptBuilder:
    def __init__(self, context, backend) -> None:
        self.context = context
        self.backend = backend

    def to_messages(self):
        return self.backend.to_messages(self.context.messages)

    def to_tools(self):
        return self.backend.to_tools(self.context.tools)

    def to_api_payload(self, *, max_output_tokens: int = 1024, tools: list | None = None) -> dict:
        return self.backend.to_payload(self.context, max_output_tokens=max_output_tokens, tools=tools)

    def parse_response(self, response: dict) -> dict:
        return self.backend.parse_response(response)

    @property
    def headers(self) -> dict:
        return self.backend.headers

    @property
    def url(self) -> str:
        return self.backend.url

File: week1_baseline/dotnet/src/Boukensha.Core/PromptBuilder.cs
using System.Text.Json.Nodes;
using Boukensha.Core.Backends;

namespace Boukensha.Core;

public sealed class PromptBuilder(Context context, ILlmBackend backend)
{
    public ILlmBackend Backend { get; } = backend;

    public JsonArray ToMessages() => Backend.ToMessages(context.Messages);

    public JsonArray ToTools() => Backend.ToTools(context.Tools);

    public JsonObject ToApiPayload(int maxOutputTokens = 1024, JsonArray? tools = null) =>
        Backend.ToPayload(context, maxOutputTokens, tools);

    public ParsedResponse ParseResponse(JsonNode response) => Backend.ParseResponse(response);

    public IReadOnlyDictionary<string, string> Headers => Backend.Headers;

    public string Url => Backend.Url;
}
